forge/src/executor/mod.rs

```diff
@@ -0,0 +1,353 @@
+        let mut account = self.db.basic(address);
+        account.balance = amount;
+
+        self.db.insert_cache(address, account);
+    }
+
+    /// Calls the `setUp()` function on a contract.
+    pub fn setup(
+        &mut self,
+        address: Address,
+    ) -> std::result::Result<(Return, Vec<RawLog>), EvmError> {
+        let CallResult { status, logs, .. } = self.call_committing::<(), _, _>(
+            Address::zero(),
+            address,
+            "setUp()",
+            (),
+            0.into(),
+            None,
+        )?;
+        Ok((status, logs))
+    }
+
+    /// Performs a call to an account on the current state of the VM.
+    ///
+    /// The state after the call is persisted.
+    pub fn call_committing<D: Detokenize, T: Tokenize, F: IntoFunction>(
+        &mut self,
+        from: Address,
+        to: Address,
+        func: F,
+        args: T,
+        value: U256,
+        abi: Option<&Abi>,
+    ) -> std::result::Result<CallResult<D>, EvmError> {
+        let func = func.into();
+        let calldata = Bytes::from(encode_function_data(&func, args)?.to_vec());
+        let RawCallResult { result, status, gas, logs, .. } =
+            self.call_raw_committing(from, to, calldata, value)?;
+        match status {
+            return_ok!() => {
+                let result = decode_function_data(&func, result, false)?;
+                Ok(CallResult { status, result, gas, logs, state_changeset: None })
+            }
+            _ => {
+                let reason = foundry_utils::decode_revert(result.as_ref(), abi)
+                    .unwrap_or_else(|_| format!("{:?}", status));
+                Err(EvmError::Execution {
+                    status,
+                    reason,
+                    gas_used: gas,
+                    logs,
+                    state_changeset: None,
+                })
+            }
+        }
+    }
+
+    /// Performs a raw call to an account on the current state of the VM.
+    ///
+    /// The state after the call is persisted.
+    pub fn call_raw_committing(
+        &mut self,
+        from: Address,
+        to: Address,
+        calldata: Bytes,
+        value: U256,
+    ) -> Result<RawCallResult> {
+        let mut evm = EVM::new();
+        evm.env = self.env.clone();
+        evm.env.tx = TxEnv {
+            caller: from,
+            transact_to: TransactTo::Call(to),
+            data: calldata,
+            value,
+            ..Default::default()
+        };
+        evm.database(&mut self.db);
+
+        // Run the call
+        let state = Rc::new(RefCell::new(ExecutorState::new()));
+        let (status, out, gas, _) = evm.inspect_commit(LogCollector::new(state.clone()));
+        let result = match out {
+            TransactOut::Call(data) => data,
+            _ => Bytes::default(),
+        };
+        let state = Rc::try_unwrap(state).expect("no inspector should be alive").into_inner();
```

Comment:
I haven't used `Rc` or `RefCell` before (only `Arc`/`Mutex`) so I'm not sure if this is a nonsense line